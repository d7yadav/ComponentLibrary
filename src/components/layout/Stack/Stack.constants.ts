import { StackDirection, StackJustify, StackAlign, StackWrap } from './Stack.types';

/**
 * STACK_DIRECTIONS component
 * 
 * @returns JSX element
 */
export const STACK_DIRECTIONS: Record<StackDirection, StackDirection> = {
  row: 'row',
  'row-reverse': 'row-reverse',
  column: 'column',
  'column-reverse': 'column-reverse',
} as const;

/**
 * STACK_JUSTIFY_CONTENT component
 * 
 * @returns JSX element
 */
export const STACK_JUSTIFY_CONTENT: Record<StackJustify, StackJustify> = {
  'flex-start': 'flex-start',
  center: 'center',
  'flex-end': 'flex-end',
  'space-between': 'space-between',
  'space-around': 'space-around',
  'space-evenly': 'space-evenly',
} as const;

/**
 * STACK_ALIGN_ITEMS component
 * 
 * @returns JSX element
 */
export const STACK_ALIGN_ITEMS: Record<StackAlign, StackAlign> = {
  'flex-start': 'flex-start',
  center: 'center',
  'flex-end': 'flex-end',
  stretch: 'stretch',
  baseline: 'baseline',
} as const;

/**
 * STACK_FLEX_WRAP component
 * 
 * @returns JSX element
 */
export const STACK_FLEX_WRAP: Record<StackWrap, StackWrap> = {
  nowrap: 'nowrap',
  wrap: 'wrap',
  'wrap-reverse': 'wrap-reverse',
} as const;

/**
 * STACK_SPACING_VALUES component
 * 
 * @returns JSX element
 */
export const STACK_SPACING_VALUES = {
  none: 0,
  xs: 0.5,
  sm: 1,
  md: 2,
  lg: 3,
  xl: 4,
  '2xl': 6,
  '3xl': 8,
} as const;

/**
 * STACK_BORDER_RADIUS component
 * 
 * @returns JSX element
 */
export const STACK_BORDER_RADIUS = {
  none: 0,
  sm: 4,
  md: 8,
  lg: 12,
  xl: 16,
  '2xl': 20,
  '3xl': 24,
  full: '50%',
} as const;

/**
 * STACK_ELEVATION_VALUES component
 * 
 * @returns JSX element
 */
export const STACK_ELEVATION_VALUES = {
  none: 0,
  sm: 1,
  md: 3,
  lg: 6,
  xl: 12,
  '2xl': 24,
} as const;

/**
 * STACK_BREAKPOINTS component
 * 
 * @returns JSX element
 */
export const STACK_BREAKPOINTS = {
  xs: 0,
  sm: 600,
  md: 900,
  lg: 1200,
  xl: 1536,
} as const;

/**
 * ACCESSIBILITY_CONSTANTS component
 * 
 * @returns JSX element
 */
export const ACCESSIBILITY_CONSTANTS = {
  listRole: 'list',
  listItemRole: 'listitem',
  groupRole: 'group',
  regionRole: 'region',
  navigationRole: 'navigation',
  toolbarRole: 'toolbar',
  menubarRole: 'menubar',
  tablistRole: 'tablist',
  defaultLabel: 'Stack layout',
} as const;

/**
 * DIVIDER_CONFIGS component
 * 
 * @returns JSX element
 */
export const DIVIDER_CONFIGS = {
  thickness: {
    thin: 1,
    medium: 2,
    thick: 4,
  },
  variants: {
    solid: 'solid',
    dashed: 'dashed',
    dotted: 'dotted',
  },
  colors: {
    divider: 'divider',
    primary: 'primary.main',
    secondary: 'secondary.main',
    error: 'error.main',
    warning: 'warning.main',
    info: 'info.main',
    success: 'success.main',
  },
} as const;

/**
 * COMMON_STACK_PATTERNS component
 * 
 * @returns JSX element
 */
export const COMMON_STACK_PATTERNS = {
  // Navigation patterns
  navbar: {
    direction: STACK_DIRECTIONS.row,
    justifyContent: STACK_JUSTIFY_CONTENT['space-between'],
    alignItems: STACK_ALIGN_ITEMS.center,
    padding: 2,
    fullWidth: true,
  },
  
  sidebar: {
    direction: STACK_DIRECTIONS.column,
    spacing: 1,
    padding: 2,
    fullHeight: true,
  },
  
  // Content patterns
  hero: {
    direction: STACK_DIRECTIONS.column,
    justifyContent: STACK_JUSTIFY_CONTENT.center,
    alignItems: STACK_ALIGN_ITEMS.center,
    spacing: 3,
    minHeight: '50vh',
    centered: true,
  },
  
  card: {
    direction: STACK_DIRECTIONS.column,
    spacing: 2,
    padding: 3,
    rounded: true,
    elevation: 2,
  },
  
  // Form patterns
  form: {
    direction: STACK_DIRECTIONS.column,
    spacing: 2,
    fullWidth: true,
  },
  
  formActions: {
    direction: STACK_DIRECTIONS.row,
    justifyContent: STACK_JUSTIFY_CONTENT['flex-end'],
    spacing: 2,
  },
  
  // Button groups
  buttonGroup: {
    direction: STACK_DIRECTIONS.row,
    spacing: 1,
  },
  
  // Lists
  list: {
    direction: STACK_DIRECTIONS.column,
    spacing: 0,
    divider: true,
  },
  
  // Media
  mediaStack: {
    direction: STACK_DIRECTIONS.column,
    spacing: 2,
    alignItems: STACK_ALIGN_ITEMS.center,
  },
} as const;

/**
 * RESPONSIVE_PATTERNS component
 * 
 * @returns JSX element
 */
export const RESPONSIVE_PATTERNS = {
  // Mobile-first responsive direction
  mobileColumn: {
    xs: STACK_DIRECTIONS.column,
    sm: STACK_DIRECTIONS.column,
    md: STACK_DIRECTIONS.row,
    lg: STACK_DIRECTIONS.row,
    xl: STACK_DIRECTIONS.row,
  },
  
  // Desktop-first responsive direction
  desktopRow: {
    xs: STACK_DIRECTIONS.column,
    sm: STACK_DIRECTIONS.row,
    md: STACK_DIRECTIONS.row,
    lg: STACK_DIRECTIONS.row,
    xl: STACK_DIRECTIONS.row,
  },
  
  // Responsive spacing
  responsiveSpacing: {
    xs: 1,
    sm: 2,
    md: 3,
    lg: 4,
    xl: 5,
  },
  
  // Responsive padding
  responsivePadding: {
    xs: 1,
    sm: 2,
    md: 3,
    lg: 4,
    xl: 6,
  },
} as const;

/**
 * DEFAULT_STACK_PROPS component
 * 
 * @returns JSX element
 */
export const DEFAULT_STACK_PROPS = {
  direction: STACK_DIRECTIONS.column,
  spacing: STACK_SPACING_VALUES.md,
  justifyContent: STACK_JUSTIFY_CONTENT['flex-start'],
  alignItems: STACK_ALIGN_ITEMS.stretch,
  flexWrap: STACK_FLEX_WRAP.nowrap,
  fullWidth: false,
  fullHeight: false,
  rounded: false,
  bordered: false,
  elevation: STACK_ELEVATION_VALUES.none,
  stretch: false,
  centered: false,
  equalWidth: false,
  equalHeight: false,
} as const;