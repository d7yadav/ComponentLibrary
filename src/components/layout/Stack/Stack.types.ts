import type { StackProps as MuiStackProps } from '@mui/material/Stack';
import type { ReactNode } from 'react';

export type StackDirection = 'row' | 'row-reverse' | 'column' | 'column-reverse';

export type StackSpacing = number | string;

export type StackJustify = 
  | 'flex-start' 
  | 'center' 
  | 'flex-end' 
  | 'space-between' 
  | 'space-around' 
  | 'space-evenly';

export type StackAlign = 
  | 'flex-start' 
  | 'center' 
  | 'flex-end' 
  | 'stretch' 
  | 'baseline';

export type StackWrap = 'nowrap' | 'wrap' | 'wrap-reverse';

export interface ResponsiveStackDirection {
  xs?: StackDirection,
  sm?: StackDirection,
  md?: StackDirection,
  lg?: StackDirection,
  xl?: StackDirection,
}

export interface ResponsiveStackSpacing {
  xs?: StackSpacing,
  sm?: StackSpacing,
  md?: StackSpacing,
  lg?: StackSpacing,
  xl?: StackSpacing,
}

export interface StackProps extends Omit<MuiStackProps, 'direction' | 'spacing' | 'justifyContent' | 'alignItems'> {
  children: ReactNode,
  
  direction?: StackDirection | ResponsiveStackDirection,
  
  spacing?: StackSpacing | ResponsiveStackSpacing,
  
  justifyContent?: StackJustify,
  
  alignItems?: StackAlign,
  
  flexWrap?: StackWrap,
  
  divider?: ReactNode,
  
  fullWidth?: boolean,
  
  fullHeight?: boolean,
  
  minHeight?: number | string,
  
  maxHeight?: number | string,
  
  minWidth?: number | string,
  
  maxWidth?: number | string,
  
  bgcolor?: string,
  
  padding?: number | string,
  
  margin?: number | string,
  
  rounded?: boolean,
  
  borderRadius?: number | string,
  
  bordered?: boolean,
  
  borderColor?: string,
  
  borderWidth?: number | string,
  
  elevation?: number,
  
  stretch?: boolean,
  
  centered?: boolean,
  
  equalWidth?: boolean,
  
  equalHeight?: boolean,
  
  gap?: number | string,
  
  rowGap?: number | string,
  
  columnGap?: number | string,
  
  className?: string,
  
  sx?: object,
  
  'aria-label'?: string,
  
  role?: string,
}

export interface StackStyleProps {
  direction: StackDirection | ResponsiveStackDirection,
  spacing: StackSpacing | ResponsiveStackSpacing,
  justifyContent: StackJustify,
  alignItems: StackAlign,
  flexWrap: StackWrap,
  fullWidth: boolean,
  fullHeight: boolean,
  minHeight?: number | string,
  maxHeight?: number | string,
  minWidth?: number | string,
  maxWidth?: number | string,
  bgcolor?: string,
  padding?: number | string,
  margin?: number | string,
  rounded: boolean,
  borderRadius?: number | string,
  bordered: boolean,
  borderColor?: string,
  borderWidth?: number | string,
  elevation: number,
  stretch: boolean,
  centered: boolean,
  equalWidth: boolean,
  equalHeight: boolean,
  gap?: number | string,
  rowGap?: number | string,
  columnGap?: number | string,
}

export interface StackDividerProps {
  orientation?: 'horizontal' | 'vertical',
  
  color?: string,
  
  thickness?: number | string,
  
  length?: number | string,
  
  variant?: 'solid' | 'dashed' | 'dotted',
}