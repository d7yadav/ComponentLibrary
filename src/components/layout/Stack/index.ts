/**
 * @fileoverview Stack component barrel export
 * @author Dilip Yadav <dilip.sm.yadav@gmail.com>
 */

export { Stack } from './Stack';
export type { 
  StackProps, 
  StackDirection, 
  StackSpacing,
  StackJustify,
  StackAlign,
  StackWrap,
  ResponsiveStackDirection,
  ResponsiveStackSpacing,
  StackStyleProps,
  StackDividerProps
} from './Stack.types';
export {
  STACK_DIRECTIONS,
  STACK_JUSTIFY_CONTENT,
  STACK_ALIGN_ITEMS,
  STACK_FLEX_WRAP,
  STACK_SPACING_VALUES,
  STACK_BORDER_RADIUS,
  STACK_ELEVATION_VALUES,
  STACK_BREAKPOINTS,
  ACCESSIBILITY_CONSTANTS,
  DIVIDER_CONFIGS,
  COMMON_STACK_PATTERNS,
  RESPONSIVE_PATTERNS,
  DEFAULT_STACK_PROPS,
} from './Stack.constants';
export {
  StackDivider,
  HStack,
  VStack,
  CenterStack,
  SpaceBetweenStack,
  CardStack,
  NavbarStack,
  SidebarStack,
  HeroStack,
  FormStack,
  ButtonGroupStack,
  ListStack,
  ResponsiveStack,
} from './Stack.styles';