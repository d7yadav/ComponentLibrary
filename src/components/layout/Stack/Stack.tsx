import { forwardRef, Children, cloneElement, isValidElement, ReactNode, ReactElement, memo } from 'react';
import { StackProps } from './Stack.types';
import { StyledStack, StackDivider } from './Stack.styles';
import {
  DEFAULT_STACK_PROPS,
  ACCESSIBILITY_CONSTANTS,
} from './Stack.constants';

/**
 * Enhanced Stack component with flexible layout and spacing control
 * 
 * Features:
 * - Flexible direction control (row, column, reverse variants)
 * - Responsive spacing and direction support
 * - Built-in divider support between items
 * - Alignment and justification controls
 * - Size constraints (width, height, min/max)
 * - Styling options: background, padding, borders, elevation
 * - Equal distribution options (width/height)
 * - Accessibility compliance (WCAG 2.1 AA)
 * - TypeScript support with strict typing
 * - Dark theme support via enhanced theme system
 */
export const Stack = forwardRef<HTMLDivElement, StackProps>((
  {
    children,
    direction = DEFAULT_STACK_PROPS.direction,
    spacing = DEFAULT_STACK_PROPS.spacing,
    justifyContent = DEFAULT_STACK_PROPS.justifyContent,
    alignItems = DEFAULT_STACK_PROPS.alignItems,
    flexWrap = DEFAULT_STACK_PROPS.flexWrap,
    divider,
    fullWidth = DEFAULT_STACK_PROPS.fullWidth,
    fullHeight = DEFAULT_STACK_PROPS.fullHeight,
    minHeight,
    maxHeight,
    minWidth,
    maxWidth,
    bgcolor,
    padding,
    margin,
    rounded = DEFAULT_STACK_PROPS.rounded,
    borderRadius,
    bordered = DEFAULT_STACK_PROPS.bordered,
    borderColor,
    borderWidth,
    elevation = DEFAULT_STACK_PROPS.elevation,
    stretch = DEFAULT_STACK_PROPS.stretch,
    centered = DEFAULT_STACK_PROPS.centered,
    equalWidth = DEFAULT_STACK_PROPS.equalWidth,
    equalHeight = DEFAULT_STACK_PROPS.equalHeight,
    gap,
    rowGap,
    columnGap,
    className,
    sx,
    'aria-label': ariaLabel,
    role,
    ...other
  },
  ref
) => {
  // Process children with dividers if provided
  const processedChildren = divider 
    ? Children.toArray(children).reduce((acc: ReactNode[], child, index) => {
        if (index > 0) {
          acc.push(
            <StackDivider
              key={`divider-${index}`}
              orientation={
                (typeof direction === 'string' ? direction : direction?.xs || 'column').includes('row') 
                  ? 'vertical' 
                  : 'horizontal'
              }
            />
          );
        }
        acc.push(child);
        return acc;
      }, [])
    : children;

  // Accessibility props
  const accessibilityProps = {
    'aria-label': ariaLabel || ACCESSIBILITY_CONSTANTS.defaultLabel,
    role: role || ACCESSIBILITY_CONSTANTS.groupRole,
  };

  return (
    <StyledStack data-testid="stack"
      ref={ref}
      direction={direction}
      spacing={spacing}
      justifyContent={justifyContent}
      alignItems={alignItems}
      flexWrap={flexWrap}
      className={className}
      sx={sx}
      // Custom props for styled component
      customFullWidth={fullWidth}
      customFullHeight={fullHeight}
      customMinHeight={minHeight}
      customMaxHeight={maxHeight}
      customMinWidth={minWidth}
      customMaxWidth={maxWidth}
      customBgcolor={bgcolor}
      customPadding={padding}
      customMargin={margin}
      customRounded={rounded}
      customBorderRadius={borderRadius}
      customBordered={bordered}
      customBorderColor={borderColor}
      customBorderWidth={borderWidth}
      customElevation={elevation}
      customStretch={stretch}
      customCentered={centered}
      customEqualWidth={equalWidth}
      customEqualHeight={equalHeight}
      customGap={gap}
      customRowGap={rowGap}
      customColumnGap={columnGap}
      {...accessibilityProps}
      {...other}
    >
      {processedChildren}
    </StyledStack>
  );
});

Stack.displayName = 'Stack';