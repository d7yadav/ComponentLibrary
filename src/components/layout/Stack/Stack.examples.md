# ðŸ“– Stack Examples

## Basic Usage

```typescript
import { Stack } from '@/components/layout/Stack';

// Basic usage
const Example = () => {
  return (
    <Stack>
      Basic Stack example
    </Stack>
  );
};
```

## Advanced Examples

```typescript
// Advanced configuration
const AdvancedExample = () => {
  const [value, setValue] = useState('');
  
  return (
    <Stack
      variant="primary"
      disabled={false}
      onChange={(e) => setValue(e.target.value)}
      sx={{
        borderRadius: 2,
        '&:hover': {
          transform: 'translateY(-2px)',
        },
      }}
    >
      Advanced Stack
    </Stack>
  );
};
```

## Integration Examples

```typescript
// Integration with form libraries
import { useController, Control } from 'react-hook-form';

interface FormIntegrationProps {
  name: string;
  control: Control;
}

const FormIntegration = ({ name, control }: FormIntegrationProps) => {
  const { field, fieldState } = useController({
    name,
    control,
    defaultValue: '',
  });

  return (
    <Stack
      {...field}
      error={!!fieldState.error}
      helperText={fieldState.error?.message}
    />
  );
};
```

## Edge Cases

```typescript
// Edge case: Empty states
const EmptyState = () => (
  <Stack disabled>
    No content available
  </Stack>
);

// Edge case: Error states
const ErrorState = () => (
  <Stack 
    error
    helperText="Something went wrong"
  />
);

// Edge case: Loading states
const LoadingState = () => (
  <Stack 
    loading
    disabled
  />
);
```

## Performance Optimization Examples

```typescript
// Memoized component for performance
const OptimizedStack = React.memo((props: StackProps) => {
  const memoizedValue = useMemo(() => {
    // Expensive calculation
    return heavyComputation(props.value);
  }, [props.value]);

  return (
    <Stack 
      {...props}
      value={memoizedValue}
    />
  );
});

// Lazy loading for large lists
const LazyList = () => {
  return (
    <VirtualizedList
      itemCount={1000}
      renderItem={({ index }) => (
        <Stack key={index}>
          Item {index}
        </Stack>
      )}
    />
  );
};
```

## Accessibility Examples

```typescript
// Full accessibility implementation
const AccessibleExample = () => {
  return (
    <Stack
      aria-label="Descriptive label for screen readers"
      aria-describedby="helper-text"
      role="button"
      tabIndex={0}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          handleAction();
        }
      }}
    >
      Accessible Stack
    </Stack>
  );
};

// High contrast mode support
const HighContrastExample = () => {
  return (
    <Stack
      sx={{
        '@media (prefers-contrast: high)': {
          borderWidth: 2,
          borderColor: 'text.primary',
        },
      }}
    >
      High contrast support
    </Stack>
  );
};
```

---

*These examples demonstrate real-world usage patterns and best practices for the Stack component.*
