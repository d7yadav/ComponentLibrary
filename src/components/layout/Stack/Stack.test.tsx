/**
 * @fileoverview Stack Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Stack } from './Stack';
import { STACK_CONSTANTS } from './Stack.constants';
import type { StackProps } from './Stack.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Stack Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<Stack>Test content</Stack>);
      expect(screen.getByRole('group')).toBeInTheDocument();
          });

    it('renders children correctly', () => {
      const testContent = 'Test Stack Content';
      renderWithTheme(<Stack>{testContent}</Stack>);
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-stack';
      renderWithTheme(
        <Stack data-testid={testId}>Test</Stack>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Stack className={customClass}>Test</Stack>
      );
      expect(screen.getByRole('group')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: StackProps['variant'][] = [
        'vertical',
        'horizontal',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Stack variant={variant} data-testid="variant-test">
              {variant} variant
            </Stack>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: StackProps['size'][] = [
        'small',
        'medium',
        'large',
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <Stack size={size} data-testid="size-test">
              {size} size
            </Stack>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });



    it('handles disabled state correctly', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Stack disabled onClick={handleClick} data-testid="disabled-test">
          Disabled Stack
        </Stack>
      );
      
      const element = screen.getByTestId('disabled-test');
      expect(element).toHaveAttribute('aria-disabled', 'true');
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('handles click events on stack container', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Stack onClick={handleClick} data-testid="clickable-stack">
          <div>Item 1</div>
          <div>Item 2</div>
        </Stack>
      );
      
      const element = screen.getByTestId('clickable-stack');
      fireEvent.click(element);
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('preserves flex direction on interaction', () => {
      renderWithTheme(
        <Stack direction="row" spacing={2} data-testid="row-stack">
          <div>Item 1</div>
          <div>Item 2</div>
        </Stack>
      );
      
      const element = screen.getByTestId('row-stack');
      expect(element).toBeInTheDocument();
      expect(element).toHaveStyle({ flexDirection: 'row' });
    });
  });



  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Stack>Accessible Stack</Stack>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <Stack aria-label={ariaLabel}>
          Stack for screen readers
        </Stack>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(<Stack>Keyboard Stack</Stack>);
      
      const element = screen.getByRole('group');
      element.focus();
      expect(element).toHaveFocus();
          });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <Stack 
          aria-describedby="description"
          aria-controls="controlled-element"
        >
          ARIA Stack
        </Stack>
      );
      
      const element = screen.getByRole('group');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiStack: {
            defaultProps: {
              variant: 'vertical',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Stack data-testid="themed-stack">
            Themed Stack
          </Stack>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-stack');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Stack>Dark Mode Stack</Stack>
        </ThemeProvider>
      );

      expect(screen.getByRole('group')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <Stack key={i}>Stack {i}</Stack>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[role="group"]')).toHaveLength(100);
    });

  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<Stack />);
      expect(screen.getByRole('group')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<Stack>{null}</Stack>);
      expect(screen.getByRole('group')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <Stack 
          variant={undefined as any} 
          size={undefined as any}
        >
          Undefined Props
        </Stack>
      );
      expect(screen.getByRole('group')).toBeInTheDocument();
    });

  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Stack>Default Stack</Stack>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: StackProps['variant'][] = [
        'vertical',
        'horizontal',
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Stack variant={variant}>{variant} Stack</Stack>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
