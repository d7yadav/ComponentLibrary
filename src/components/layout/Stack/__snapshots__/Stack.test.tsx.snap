// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Stack Component > Snapshots > matches snapshot for all variants > variant-horizontal 1`] = `
<div
  aria-label="Stack layout"
  class="MuiStack-root css-i6qcix-MuiStack-root"
  data-testid="stack"
  role="group"
  variant="horizontal"
>
  horizontal
   Stack
</div>
`;

exports[`Stack Component > Snapshots > matches snapshot for all variants > variant-vertical 1`] = `
<div
  aria-label="Stack layout"
  class="MuiStack-root css-i6qcix-MuiStack-root"
  data-testid="stack"
  role="group"
  variant="vertical"
>
  vertical
   Stack
</div>
`;

exports[`Stack Component > Snapshots > matches snapshot for default props 1`] = `
<div
  aria-label="Stack layout"
  class="MuiStack-root css-i6qcix-MuiStack-root"
  data-testid="stack"
  role="group"
>
  Default Stack
</div>
`;
